text/0126-introduce-xcq.md

```diff
@@ -0,0 +1,342 @@
+
+- Exported Functions:
+   - `main`: The entry point of the XCQ program. It performs type checking, arguments and results passing and executes the query.
+
+### XCQ Program Execution Flow
+
+The interaction between an XCQ program and the XCQ Extension Executor follows these steps:
+
+1. Program Loading: The Executor loads the PolkaVM program binary.
+
+2. Environment Setup: The Executor configures the PolkaVM environment, registering host functions like `host_call` and `return_ty`.
+
+3. Main Function Execution: The Executor calls the program's `main` function, passing serialized query arguments.
+
+4. Program Execution:
+   1. Type Checking: The program uses the `return_ty` function to ensure compatibility with supported chain extensions.
+   2. Query Execution: The program executes the query using `host_call` and performs custom computations.
+   3. Result Serialization: The program serializes the result, writes it to shared memory, and returns the pointer and length to the executor.
+
+5. Result Retrieval: The Executor reads the result from shared memory and returns it to the caller.
+
+## Drawbacks
+
+### Performance issues
+
+- XCQ Query Program Size: The size of XCQ query programs should be optimized to ensure efficient storage and transmission via XCMP/HRMP.
+Some strategies to address this issue include:
+  - Exploring modular program structures that allow for separate storage and transmission of core logic and supporting elements. PolkaVM supports spliting the program into multiple modules.
+  - Establishing guidelines for optimizing dynamic memory usage within query programs
+
+### User experience issues
+
+- Debugging: Currently, there is no full-fledged debuggers for PolkaVM programs. The only debugging approach is to set the PolkaVM backend in interpreter mode and then log the operations at the assembly level, which is too low-level to debug efficiently.
+- Gas computation: According to [this issue](https://github.com/koute/polkavm/issues/17), the gas cost model of PolkaVM is not accurate for now.
+
+## Testing, Security, and Privacy
+
+- Testing:
+  - A comprehensive test suite should be developed to cover various scenarios:
+    - Positive test cases:
+      - Basic queries with various extensions, data types, return values, custom computations, etc.
+      - Accurate conversion between given weight limit and the gas limit of PolkaVM
+    - Negative test cases:
+      - Queries exceeding weight limits
+      - Invoking queries from unauthorized sources
+    - Edge cases:
+      - Queries with minimal or maximum allowed input sizes
+  - Integration tests to ensure proper interaction with off-chain wallets/UI and on-chain XCM, including the aforementioned use cases in **Motivation** section.
+
+- Security:
+  - The XCQ system must enforce a strict read-only policy for all query operations. A mechanism should be implemented to prevent any state-changing operations within XCQ queries.
```

Comment:
Could the execution of the program be sandboxed?

Reply:
Yes, we can do the query in a transaction, and rollback finally.